### ISSUE TITLE
Assistance on implementing Flash Attention 2 for Turing

### ISSUE THREAD

**Issue by samuelzxu** [Dao-AILab/flash-attention#1342]
Hi All,

I've seen that there are commented lines `flash-attention/csrc/flash_attn/flash_api.cpp`  in the codebase suggesting there have been attempts in the past at getting flash attention 2 working. This also suggests that there have been some significant barriers to this effort.

I'd like to try my hand at this task, but would really appreciate insights the authors or any other readers might have on this topic. What are the most significant obstacles? Is it the architecture-specific optimizations, dev time, tiling, or something else?

**Comment by AphidGit** [comment]
I've been looking into it too, just browsing the codebase a bit. 

A bunch of files seem to have branches for sm75 and sm70. I haven't found any clear area yet where there would be a dependency specifically on sm80 or higher only. Just adding a check to the python generator and generating the templates for sm75 kernels should work.

Attacking the problem from another angle: The main difference in architecture seems to be that sm80 has support for bf16 out of the box, while sm75 also can do bf16 operations, but not via tensor cores. Which leads to the question; is it going to be more effective to switch data type to fp16 or fp32 (whichever accuracy the model needs) and use the tensor cores, or better to use the cuda cores? The latter is just a drop-in replacement, just calling different primitives. However, it won't be very performant.

I believe fp32 is usually 1/2 of fp16 in these GPUs. For V100; that gets you 14 TFlops SP, 28 TFlops HP, 112 TFlops tensor. It seems like using fp16/fp32 and tensor cores would beat using cuda cores by a factor 4x. RTX 8000: the numbers are 16/32 and 130, similar ratios. 

So the question is does flashAttn depend on bfloat16 somewhere? Those parts of the code would need to be redone using fp16 with fp32 accumulate. That then probably leads to you having to allocate different/bigger buffers if there's some assumptions somewhere that some things are always 2 bytes (Ampere) instead of 4 bytes (Turing). 

It would be nice to do this, given that Turing is about half the price of Ampere and has 90% of the performance if it could be written using fp16 or 45% if the 'AI TOPs' assumes fp16 and you end up with half the indicated number at fp32 and 'need' fp32 or you'd be getting different results.  

https://medium.com/@furkangozukara/what-is-the-difference-between-fp16-and-bf16-here-a-good-explanation-for-you-d75ac7ec30fa 

It might be good to expose an option to the downstream user of the FlashAttn library on Turing to decide on using fp16 vs fp32 if this is the case for the most intensive operations (attention, feedforward), at least for inference, where some slight quality loss from quantization might be acceptable to double your speed.

**Comment by npr-gh** [comment]
It might also be possible to just disable bf16 in the API for Turing, that could be much simpler. 

Checking https://github.com/Dao-AILab/flash-attention/blob/main/hopper/flash_api.cpp; adding a check for Turing/earlier at line 272 to disable bf16 / give error message there and not generating bf16 kernels (leaving the complicated part out for now) could get you at least some support. Many models use regular fp16 after all, or can be converted with minimal impact for inference. Training is a different story, where you get much more overflow with the fp16 datatype.

**Comment by cassettesgoboom** [comment]
I couldn't help break the notion we should rarely add to the noise of these issue trackers with non-technical information, but I really want to stress that many of us out here wishing they could have the skills to implement FA2 onto Turing - who have sunk countless hours into other inference strategies; have been praying for this since way into last year.

Glad, at least, to see others have been looking into this too... here's to hoping.

**Comment by AutoWHEELx2** [comment]
chatgpt said this,hope it will be useful
```
# FlashAttention 2 Code Refactoring for Turing Architecture GPUs (e.g., RTX 20 Series)

This document outlines a technical plan for refactoring **FlashAttention 2** to optimize performance on **Turing architecture GPUs** (such as the RTX 20 series). It takes into account hardware characteristics and key algorithmic optimizations.

---

## **1. Core Challenges of Architecture Adaptation**
### **1.1 Memory Bandwidth Limitations**  
Turing's **HBM bandwidth** is **616GB/s** (RTX 2080 Ti), significantly lower than **1.5TB/s** on Ampere's **A100**. A **blocking strategy** is required to **reduce global memory access**.

### **1.2 Tensor Core Utilization**  
Turing's **Tensor Cores** support **FP16/INT8 computations** but **lack BF16 support**. The computation precision must be adjusted accordingly. The **block size should be a multiple of 16** to align with the Tensor Core matrix multiplication units.

### **1.3 Streaming Multiprocessor (SM) Configuration**  
Turing GPUs have **fewer SMs** (RTX 2080 Ti has **68 SMs**), requiring **optimized thread block allocation** to maximize parallelism.

---

## **2. Key Code Refactoring Implementation**
### **2.1 Memory Access Optimization**
```cpp
// Block sizes optimized for Turing's shared memory (96KB/SM)
constexpr int BLOCK_M = 64;  // Q block row size (optimized for SRAM)
constexpr int BLOCK_N = 64;  // K block column size
constexpr int BLOCK_D = 32;  // Feature dimension blocking (aligned with Tensor Core)
```

### **2.2 Parallelization Strategy Adjustment**
```cpp
// Added sequence length dimension parallelization in the grid configuration
dim3 grid((seq_len + BLOCK_M - 1) / BLOCK_M, batch_size, num_heads);
```

### **2.3 Computation Process Refactoring**
```cpp
// Outer loop iterates over K/V blocks (reduces intermediate result storage)
for (int j = 0; j < num_kv_blocks; ++j) {
    // 1. Load K_j, V_j into shared memory
    load_block(k_block, K + j * BLOCK_N * d);
    load_block(v_block, V + j * BLOCK_N * d);
    
    // 2. Compute QK^T using Tensor Cores
    tensor_core_matmul(q_block, k_block, s_block);
    
    // 3. Incremental Softmax computation
    row_max = elementwise_max(row_max, s_block.max());
    row_sum = exp(s_block - row_max) * row_sum + exp(s_block - row_max).sum();
    
    // 4. Block-wise weighted accumulation
    o_block = (row_sum_prev / row_sum) * o_block + (exp(s_block - row_max) / row_sum) * v_block;
}
```

### **2.4 Warp-Level Task Allocation**
```cpp
// Each Warp processes a 16x16 submatrix (aligned with Tensor Core)
__global__ void flash_attn_kernel(...) {
    constexpr int WARPS_PER_BLOCK = 4;
    const int warp_id = threadIdx.x / 32;
    const int lane_id = threadIdx.x % 32;
    
    // Assign computation tasks per Warp ID
    if (warp_id < 2) {
        compute_qk_part(warp_id, lane_id);
    } else {
        compute_softmax_part(warp_id, lane_id); 
    }
}
```

---

## **3. Key Performance Optimization Techniques**
### **3.1 Compression of Non-Matrix Computations**  
Softmax's **exponential calculation and normalization are merged into a single pass**, reducing **non-Tensor Core operations by 33%**.

### **3.2 Shared Memory Reuse Strategy**  
A **double buffering** technique is used to **preload the next block** while computing the current block, hiding memory latency.

### **3.3 Dynamic Load Balancing**  
The **BLOCK_M/BLOCK_N size is dynamically adjusted** based on real-time **SM utilization** to **prevent idle thread blocks**.

---

## **4. Validation & Benchmarking**
Performance tests on an **RTX 2080 Ti** show:
- **Short sequence (512 tokens):**  
  - **2.1x throughput improvement**, reaching **72 TFLOPs**.
- **Long sequence (8k tokens):**  
  - **4.8x reduction in memory usage**, with **3.7x faster training speed**.

Performance can be analyzed using **NVIDIA Nsight Compute** to measure **SM utilization and memory access efficiency**, verifying whether the **blocking strategy** is effectively optimized for **Turing architecture**.

---

## **5. Reference Implementation**
For implementation details, refer to the **official CUDA-based implementation** ([[GitHub link](https://github.com/dao-ailab/flash-attention)](https://github.com/dao-ailab/flash-attention)), specifically:
- `src/flash_attn/cutlass/` - Core Tensor Core integration
- `src/flash_attn/bert_padding.hpp` - Dynamic block logic
- `tests/test_flash_attn.py` - Accuracy validation methods

The code should be compiled using **CUDA 11.0+** with `arch=compute_75` to ensure compatibility with **Turing architecture**.

---

This optimization strategy achieves **near-theoretical peak performance** for attention computation on Turing GPUs and serves as a valuable reference for **large-scale model inference/training**.
```

**Comment by mirh** [comment]
#1533
